### ISSUE TITLE
DefaultGradient setGradientFor review functionality for revision

### ISSUE THREAD

**Issue by nyghtowl** [deeplearning4j/deeplearning4j#1981]
tl;dr: not a blocker for gradient calculation, but review setGradientFor to see if functionality can align more to method name.

Currently setGradientFor updates the gradient map but when gradient() is called on the object it pulls from the INDArray flattenedGradient which is not updated. Should setGradientFor update the object fully so however it's referenced, the updated gradients will be provided by the object?

Noting this as an issue to revisit. The main problem with updating flattenedGradient in this method  is it causes CheckGradientTests to fail which seems to be due to conflicts in how backprop uses layer gradient views for the calculation.
